packages/swing-store/src/swingStore.js

```diff
@@ -623,6 +835,53 @@ export function initSwingStore(dirPath = null, options = {}) {
   return makeSwingStore(dirPath, true, options);
 }
 
+/**
+ * @param {SwingStoreExporter} exporter
+ * @param {string | null} [dirPath]
+ * @param {object} options
+ * @returns {Promise<SwingStore>}
+ */
+export async function importSwingStore(exporter, dirPath = null, options = {}) {
+  if (dirPath) {
+    assert.typeof(dirPath, 'string');
+  }
+  const { includeHistorical = false } = options;
+  const store = makeSwingStore(dirPath, true, options);
+  const { kernelStorage } = store;
+  const artifactMetadata = new Map();
+  for (const [key, value] of exporter.getKVData()) {
+    if (key.startsWith('export.')) {
+      artifactMetadata.set(key, value);
+    } else if (value === null) {
+      kernelStorage.kvStore.delete(key);
+    } else {
+      kernelStorage.kvStore.set(key, value);
+    }
+  }
+  for await (const artifactName of exporter.getArtifactNames(
+    includeHistorical,
+  )) {
```

Comment:
I think both are valid ways to do it. They differ on how the importer needs to manage artifacts.

In the push approach, the importer should finish the first phase (export data / IAVL pairs) with a list of every artifact that it must receive before the import is complete, and check them off as they arrive. It might receive artifacts that it doesn't want (which it should ignore?). If we ever add intermediate artifacts (e.g. one artifact that holds a linear list of span hashes for a single vat, the effective parent of a set of one-per-span artifacts), where the export-data/IAVL-pairs are the parent of a hash tree, then the writer might push a grandchild node before the intermediate node, and the importer would need to buffer things or build up a list of "things that still need validation" to avoid finishing with unvalidated data. (Lots of Merkle-tree -based security schemes have broken this way).

In the pull approach, it's easier to request hash parents first, and only request children after the importer is prepared to validate them. But it requires that the data *provider* have random access to the artifacts, which may or may not be convenient. In a fully streaming / zero-copy system, you don't get to buffer in the middle to accomodate ordering differences between the writer and the reader. And it might be inconvenient for the writer to order things in a way that lets the reader/importer avoid these forward validation references.

I kind of like the pull approach better, but I like @FUDCo 's elegance in letting the export API be plugged directly into the import API, so I guess I don't have a strong opinion.

Reply:
> it requires that the data _provider_ have random access to the artifacts, which may or may not be convenient.

The Exporter API is currently design to effectively require this random access through `getArtifact()`. I'm ok with it. We can always evolve the API later if we want to support streaming of the whole export with zero copy.

> I like @FUDCo 's elegance in letting the export API be plugged directly into the import API, so I guess I don't have a strong opinion.

Yeah I designed the API so that the exporter interface could be used directly by the importer, which is why it supports a push of names and a pull of artifacts themselves. I figured the importer would likely only use the later, ignoring the explicit names enumeration.